python/ray/projects/scripts.py

```diff
@@ -110,3 +117,130 @@ def create(project_name, cluster_yaml, requirements):
+    attach_cluster(
+        project_definition["cluster"],
+        start=False,
+        use_tmux=False,
+        override_cluster_name=None,
+        new=False,
+    )
+
+
+@session_cli.command(help="Stop a session based on current project config")
+def stop():
+    project_definition = load_project_or_throw()
+    teardown_cluster(
+        project_definition["cluster"],
+        yes=True,
+        workers_only=False,
+        override_cluster_name=None)
+
+
+@session_cli.command(help="Start a session based on current project config")
+def start():
+    project_definition = load_project_or_throw()
+
+    cluster_yaml = project_definition["cluster"]
+    working_directory = project_definition["name"]
+
+    logger.info("[1/4] Creating cluster")
+    create_or_update_cluster(
+        config_file=cluster_yaml,
+        override_min_workers=None,
+        override_max_workers=None,
+        no_restart=False,
+        restart_only=False,
+        yes=True,
+        override_cluster_name=None,
+    )
+
+    logger.info("[2/4] Syncing the repo")
+    if "repo" in project_definition:
+        # HACK: Skip git clone if exists so the this command can be idempotent
+        # More advanced repo update behavior can be found at
+        # https://github.com/jupyterhub/nbgitpuller/blob/master/nbgitpuller/pull.py
+        session_exec_cluster(
+            cluster_yaml,
+            "git clone {repo} {directory} || true".format(
+                repo=project_definition["repo"],
+                directory=project_definition["name"]),
+            cwd=working_directory,
+        )
+
+    logger.info("[3/4] Setting up environment")
+    _setup_environment(
+        cluster_yaml, project_definition["environment"], cwd=working_directory)
+
+    logger.info("[4/4] Running commands")
+    _run_commands(
+        cluster_yaml, project_definition["commands"], cwd=working_directory)
+
+
+def session_exec_cluster(cluster_yaml, cmd, cwd):
+    cmd = "cd {cwd}; {cmd}".format(cwd=cwd, cmd=cmd)
+    exec_cluster(
+        config_file=cluster_yaml,
+        cmd=cmd,
+        docker=False,
+        screen=False,
+        tmux=False,
+        stop=False,
+        start=False,
+        override_cluster_name=None,
+        port_forward=None,
+    )
+
+
+def _setup_environment(cluster_yaml, project_environment, cwd):
+
+    need_docker = ("dockerfile" in project_environment
+                   or "dockerimage" in project_environment)
+    if need_docker:
+        raise NotImplementedError(
+            "Docker support in session is currently not implemented."
+            "Please file an feature request at"
+            "https://github.com/ray-project/ray/issues")
+
+    if "requirements" in project_environment:
+        requirements_txt = project_environment["requirements"]
+
+        # Create a temporary requirements_txt in the head machine
+        remote_requirements_txt = (
+            "/tmp/" + "ray_project_requirements_txt_{}".format(time.time()))
```

Comment:
tempfile?

Reply:
This is supposed to be a file in the remote headnode. So I don't think we can use temp file.